File: Sources/VoxLocalCore/Transcription/WhisperTranscriber.swift
import Foundation

/// Runs the bundled whisper-cli against a WAV file and returns parsed text.
public final class WhisperTranscriber: @unchecked Sendable {
    public struct Transcript: Equatable, Sendable {
        public let text: String
        public let detectedLanguage: String?
    }

    private let runner: SubprocessRunning
    private let binaryOverride: URL?

    public init(runner: SubprocessRunning = ProcessSubprocessRunner(), binaryOverride: URL? = nil) {
        self.runner = runner
        self.binaryOverride = binaryOverride
    }

    /// Candidate locations for whisper-cli, in priority order:
    /// 1. `VOXLOCAL_WHISPER_CLI` environment variable (tests, development),
    /// 2. the app bundle's auxiliary executable (packaged app),
    /// 3. the in-repo build output (running via `swift run` from the repo).
    public static func binaryCandidates() -> [URL] {
        var candidates: [URL] = []
        if let env = ProcessInfo.processInfo.environment["VOXLOCAL_WHISPER_CLI"], !env.isEmpty {
            candidates.append(URL(fileURLWithPath: env))
        }
        if let bundled = Bundle.main.url(forAuxiliaryExecutable: "whisper-cli") {
            candidates.append(bundled)
        }
        let cwd = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        candidates.append(cwd.appendingPathComponent("vendor/whisper.cpp/build/bin/whisper-cli"))
        return candidates
    }

    public func locateBinary() throws -> URL {
        if let binaryOverride {
            return binaryOverride
        }
        let candidates = Self.binaryCandidates()
        let fm = FileManager.default
        for url in candidates where fm.isExecutableFile(atPath: url.path) {
            return url
        }
        let searched = candidates.map { PathRedactor.redact($0.path) }.joined(separator: ", ")
        throw VoxLocalError.whisperBinaryMissing(searched)
    }

    public func transcribe(
        audioURL: URL,
        modelURL: URL,
        language: String,
        threads: Int,
        removeArtifacts: Bool,
        timeout: TimeInterval = 180
    ) async throws -> Transcript {
        let binary = try locateBinary()

        guard FileManager.default.fileExists(atPath: audioURL.path) else {
            throw VoxLocalError.recordingFailed("audio file vanished before transcription")
        }
        switch ModelManager.validateModelFile(at: modelURL) {
        case .valid: break
        case .missing: throw VoxLocalError.modelMissing(modelURL.path)
        case .tooSmall, .badMagic: throw VoxLocalError.modelInvalid(modelURL.path)
        }

        let outputBase = FileManager.default.temporaryDirectory
            .appendingPathComponent("voxlocal-out-\(UUID().uuidString)")
        let jsonURL = URL(fileURLWithPath: outputBase.path + ".json")
        defer { try? FileManager.default.removeItem(at: jsonURL) }

        let args = WhisperCommandBuilder.arguments(
            modelPath: modelURL.path,
            audioPath: audioURL.path,
            language: language,
            threads: threads,
            outputBase: outputBase.path)

        Log.shared.info("whisper-cli starting (model \(modelURL.lastPathComponent), lang \(language), threads \(threads))")
        let started = Date()

        let result: SubprocessResult
        do {
            result = try await runner.run(executable: binary, arguments: args, timeout: timeout)
        } catch SubprocessError.timeout {
            throw VoxLocalError.transcriptionTimeout
        } catch SubprocessError.launchFailed(let reason) {
            if reason.localizedCaseInsensitiveContains("bad cpu type")
                || reason.localizedCaseInsensitiveContains("badarch") {
                throw VoxLocalError.unsupportedArchitecture(reason)
            }
            throw VoxLocalError.transcriptionFailed(exitCode: -1, detail: reason)
        }

        try Task.checkCancellation()

        guard result.exitCode == 0 else {
            let stderrTail = String(data: result.stderr.suffix(600), encoding: .utf8) ?? ""
            if stderrTail.localizedCaseInsensitiveContains("failed to load model")
                || stderrTail.localizedCaseInsensitiveContains("invalid model") {
                throw VoxLocalError.modelInvalid(modelURL.path)
            }
            throw VoxLocalError.transcriptionFailed(exitCode: result.exitCode, detail: stderrTail)
        }

        guard let jsonData = try? Data(contentsOf: jsonURL) else {
            throw VoxLocalError.transcriptionOutputMissing
        }
        let output: WhisperOutputParser.Output
        do {
            output = try WhisperOutputParser.parse(jsonData: jsonData, removeArtifacts: removeArtifacts)
        } catch {
            throw VoxLocalError.transcriptionOutputMissing
        }

        let elapsed = String(format: "%.1f", Date().timeIntervalSince(started))
        Log.shared.info("whisper-cli finished in \(elapsed)s (chars: \(output.text.count), lang: \(output.detectedLanguage ?? "?"))")
        return Transcript(text: output.text, detectedLanguage: output.detectedLanguage)
    }
}

File: windows/src/VoxLocal.App/Core/Transcription/WhisperTranscriber.cs
using System.Text;
using VoxLocal.Core.Models;
using VoxLocal.Core.Utilities;

namespace VoxLocal.Core.Transcription;

/// <summary>Runs the bundled whisper-cli.exe against a WAV file and returns parsed text.</summary>
public sealed class WhisperTranscriber
{
    public sealed record Transcript(string Text, string? DetectedLanguage);

    private readonly ISubprocessRunner _runner;
    private readonly string? _binaryOverride;

    public WhisperTranscriber(ISubprocessRunner? runner = null, string? binaryOverride = null)
    {
        _runner = runner ?? new ProcessSubprocessRunner();
        _binaryOverride = binaryOverride;
    }

    /// <summary>
    /// Candidate locations for whisper-cli.exe, in priority order:
    /// 1. VOXLOCAL_WHISPER_CLI environment variable (tests, development),
    /// 2. next to the application executable (packaged app),
    /// 3. the in-repo CMake build output (running from the repo).
    /// </summary>
    public static IReadOnlyList<string> BinaryCandidates()
    {
        var candidates = new List<string>();
        var env = Environment.GetEnvironmentVariable("VOXLOCAL_WHISPER_CLI");
        if (!string.IsNullOrEmpty(env))
        {
            candidates.Add(env);
        }
        candidates.Add(Path.Combine(AppContext.BaseDirectory, "whisper-cli.exe"));
        var cwd = Directory.GetCurrentDirectory();
        candidates.Add(Path.Combine(cwd, "vendor", "whisper.cpp", "build", "bin", "Release", "whisper-cli.exe"));
        candidates.Add(Path.Combine(cwd, "vendor", "whisper.cpp", "build", "bin", "whisper-cli.exe"));
        return candidates;
    }

    public string LocateBinary()
    {
        if (_binaryOverride is not null)
        {
            return _binaryOverride;
        }
        var candidates = BinaryCandidates();
        foreach (var path in candidates)
        {
            // Windows has no executable bit; existence is the practical check.
            if (File.Exists(path)) return path;
        }
        var searched = string.Join(", ", candidates.Select(PathRedactor.Redact));
        throw new VoxLocalException(new VoxLocalError.WhisperBinaryMissing(searched));
    }

    /// <summary>Non-throwing variant of <see cref="LocateBinary"/> for status screens (onboarding).</summary>
    public bool TryLocateBinary(out string? path)
    {
        try
        {
            path = LocateBinary();
            return true;
        }
        catch (VoxLocalException)
        {
            path = null;
            return false;
        }
    }

    public async Task<Transcript> TranscribeAsync(
        string audioPath,
        string modelPath,
        string language,
        int threads,
        bool removeArtifacts,
        TimeSpan? timeout = null,
        CancellationToken cancellationToken = default)
    {
        var binary = LocateBinary();

        if (!File.Exists(audioPath))
        {
            throw new VoxLocalException(new VoxLocalError.RecordingFailed("audio file vanished before transcription"));
        }
        switch (ModelManager.ValidateModelFile(modelPath))
        {
            case ModelManager.ValidationResult.Valid:
                break;
            case ModelManager.ValidationResult.Missing:
                throw new VoxLocalException(new VoxLocalError.ModelMissing(modelPath));
            default:
                throw new VoxLocalException(new VoxLocalError.ModelInvalid(modelPath));
        }

        var outputBase = Path.Combine(Path.GetTempPath(), $"voxlocal-out-{Guid.NewGuid()}");
        var jsonPath = outputBase + ".json";
        try
        {
            var args = WhisperCommandBuilder.Arguments(modelPath, audioPath, language, threads, outputBase);

            Log.Shared.Info($"whisper-cli starting (model {Path.GetFileName(modelPath)}, lang {language}, threads {threads})");
            var started = DateTime.UtcNow;

            SubprocessResult result;
            try
            {
                result = await _runner.RunAsync(binary, args, timeout ?? TimeSpan.FromSeconds(180), cancellationToken)
                    .ConfigureAwait(false);
            }
            catch (SubprocessTimeoutException)
            {
                throw new VoxLocalException(new VoxLocalError.TranscriptionTimeout());
            }
            catch (SubprocessLaunchException ex)
            {
                // ERROR_BAD_EXE_FORMAT (193): wrong bitness or corrupted exe —
                // the Windows analogue of macOS "bad CPU type".
                if (ex.InnerException is System.ComponentModel.Win32Exception { NativeErrorCode: 193 }
                    || ex.Message.Contains("not a valid", StringComparison.OrdinalIgnoreCase))
                {
                    throw new VoxLocalException(new VoxLocalError.UnsupportedArchitecture(ex.Message));
                }
                throw new VoxLocalException(new VoxLocalError.TranscriptionFailed(-1, ex.Message));
            }

            cancellationToken.ThrowIfCancellationRequested();

            if (result.ExitCode != 0)
            {
                var tail = result.Stderr.Length <= 600 ? result.Stderr : result.Stderr[^600..];
                var stderrTail = Encoding.UTF8.GetString(tail);
                if (stderrTail.Contains("failed to load model", StringComparison.OrdinalIgnoreCase)
                    || stderrTail.Contains("invalid model", StringComparison.OrdinalIgnoreCase))
                {
                    throw new VoxLocalException(new VoxLocalError.ModelInvalid(modelPath));
                }
                throw new VoxLocalException(new VoxLocalError.TranscriptionFailed(result.ExitCode, stderrTail));
            }

            byte[] jsonData;
            try
            {
                jsonData = await File.ReadAllBytesAsync(jsonPath, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException)
            {
                throw new VoxLocalException(new VoxLocalError.TranscriptionOutputMissing());
            }

            WhisperOutputParser.Output output;
            try
            {
                output = WhisperOutputParser.Parse(jsonData, removeArtifacts);
            }
            catch (Exception)
            {
                throw new VoxLocalException(new VoxLocalError.TranscriptionOutputMissing());
            }

            var elapsed = (DateTime.UtcNow - started).TotalSeconds;
            Log.Shared.Info($"whisper-cli finished in {elapsed:F1}s (chars: {output.Text.Length}, lang: {output.DetectedLanguage ?? "?"})");
            return new Transcript(output.Text, output.DetectedLanguage);
        }
        finally
        {
            try { File.Delete(jsonPath); } catch { /* best effort */ }
        }
    }
}
